Add verification for primitive objects

ehrenjulzert opened the pull request:
For issue #14741

- Add checks for invalid referencetype field/method descriptors

- Add new error messages to cfrerr.nls

- Fix cases in staticverify.c that are preventing value types tests
  from running with Xverify enabled

Signed-off-by: Ehren Julien-Neitzert <ehren.julien-neitzert@ibm.com>

ChengJin01 commented on runtime/bcutil/cfreader.c:
```diff
@@ -2063,6 +2065,14 @@ checkAttributes(J9PortLibrary* portLib, J9CfrClassFile* classfile, J9CfrAttribut
 					goto _errorFound;
 				}
+#if defined(J9VM_OPT_VALHALLA_VALUE_TYPES)
+				if (cpBase[value].tag == CFR_CONSTANT_Class) {
```
`CFR_CONSTANT_Class` should be placed on the left side when using `==`.

ChengJin01 commented:
FYI: @tajila, @DanHeidinga

ehrenjulzert commented on runtime/verutil/chverify.c:
```diff
@@ -212,3 +212,29 @@ I_32
+		result = FALSE;
+	} else if (
+		('[' == bytes[0]) ||
+		((';' == bytes[length - 1]) && (('L' == bytes[0]) || ('Q' == bytes[0])))
```
Ok, I've addressed all your comments in 38d0302

hangshao0 commented on runtime/bcverify/vrfyhelp.c:
```diff
@@ -256,8 +256,16 @@ buildStackFromMethodSignature( J9BytecodeVerificationData *verifyData, UDATA **s
 			/* This is <init>, not java/lang/Object */
-			PUSH(BCV_SPECIAL_INIT | (classIndex << BCV_CLASS_INDEX_SHIFT));
+#if defined(J9VM_OPT_VALHALLA_VALUE_TYPES)
+			if (J9_ARE_ALL_BITS_SET(romClass->modifiers, J9AccPrimitiveValueType)) { /* This is temporary and will have to be changed when <vnew> is introduced */
```
Please use `J9ROMCLASS_IS_PRIMITIVE_VALUE_TYPE(romClass)`, and also at least mention `J9VM_OPT_VALHALLA_NEW_FACTORY_METHOD` in the comment.

hangshao0 commented on runtime/oti/bytecodewalk.h:
```diff
@@ -102,6 +102,9 @@ base types: (in the 19bit class index field)
 #define BCV_SPECIAL_INIT					4		/* set bit means special init object ("this" for <init>) */
 #define BCV_SPECIAL_NEW						8		/* set bit means special new object (PC offset in upper 28 bits) */
+#if defined(J9VM_OPT_VALHALLA_VALUE_TYPES)
+#define BCV_PRIMITIVE_CLASS					16		/* set bit means primitive class */
```
The flag is `J9AccPrimitiveValueType`, so I suggest to use `BCV_PRIMITIVE_VALUETYPE`.

hangshao0 commented on runtime/oti/j9.h:
```diff
@@ -350,7 +350,6 @@ static const struct { \
 #define IS_LTYPE(firstChar) ('L' == (firstChar))
 
 #define J9_IS_STRING_DESCRIPTOR(str, strLen) (((strLen) > 2) && (IS_REF_OR_VAL_SIGNATURE(*(str))) && (';' == *((str) + (strLen) - 1)))
-
```
No need to include this file in this PR, only an empty line change in this file.

ehrenjulzert replied:
I'm not using `IS_REF_OR_VAL_SIGNATURE` because it only checks the first character of the string that is passed to it. `bcvIsReferenceTypeDescriptor` is used to check whether a string is a normal class name (like `MyClass`)  or an equivalent reference type descriptor (like `LMyClass;`). If I only checked the first character then class names that began with L or Q would be seen as reference type descriptors even if they were not.

hangshao0 commented on runtime/verutil/chverify.c:
```diff
@@ -212,3 +212,30 @@ I_32
+	U_32 length = info->slot1;
+	U_8 *bytes = info->bytes;
+	BOOLEAN result = FALSE;
+
```
There are existing macros `IS_REF_OR_VAL_SIGNATURE()` and `IS_REF_OR_VAL_SIGNATURE()` you may use in this function.

ehrenjulzert commented on runtime/bcutil/cfreader.c:
```diff
@@ -2233,17 +2258,23 @@ checkAttributes(J9PortLibrary* portLib, J9CfrClassFile* classfile, J9CfrAttribut
+					if (
+#if defined(J9VM_OPT_VALHALLA_VALUE_TYPES)
+						bcvIsReferenceTypeDescriptor(classInfoUtf8)
+#else /* #if defined(J9VM_OPT_VALHALLA_VALUE_TYPES) */
```
In non-valhalla builds we just want to check if `outer_class_info_index` points to an array type, but in Valhalla builds we want to check if it points to an array type, Q type, or L type (all of which are checked by `bcvIsReferenceTypeDescriptor`)

hangshao0 replied:
Ah, OK

ehrenjulzert commented:
@hangshao0 I have addressed your previous comments in this new commit
I also fixed the errors with QTypes being assignable to null by adding more qtype checks in verifyhelper.c, and by changing errormessageframeworkrtv.c/errormessagehelper.c to ensure the error is being printed correctly

hangshao0 commented on runtime/nls/cfre/cfrerr.nls:
```diff
@@ -1590,3 +1590,79 @@ J9NLS_CFR_ERR_PREVIEW_VERSION2.explanation=Classfiles with minor version 65535 m
 J9NLS_CFR_ERR_PREVIEW_VERSION2.user_response=Run with the correct Java runtime version and use --enable-preview, or re-compile the code with the same version of the runtime.
 # END NON-TRANSLATABLE
+
+J9NLS_INVALID_REFERENCETYPE_DESCRIPTOR_ON_CLASS=A ReferenceType descriptor cannot be referenced by a CONSTANT_Class_info structure that represents this_class
```
I see there are 3 ways of saying "reference type" in this file:

"ReferenceType", "referencetype" and "reference type". Better to be consistent.
Better to use "reference type" if you want it to be translated. Or use "ReferenceType" and add `#  ReferenceType is not translatable` above the message.

hangshao0 replied:
Please add `# this_class is not translatable`, same for super_class in the below message.

hangshao0 commented on runtime/nls/cfre/cfrerr.nls:
```diff
@@ -1590,3 +1590,79 @@ J9NLS_CFR_ERR_PREVIEW_VERSION2.explanation=Classfiles with minor version 65535 m
+J9NLS_INVALID_REFERENCETYPE_DESCRIPTOR_ON_SUPER_CLASS.user_response=Contact the provider of the class file for a corrected version.
+# END NON-TRANSLATABLE
+
+J9NLS_INVALID_REFERENCETYPE_DESCRIPTOR_ON_INTERFACE=A ReferenceType descriptor cannot be referenced by a CONSTANT_Class_info structure that represents the interfaces array
```
"interfaces array" --> "an entry in the interface array".

hangshao0 commented on runtime/nls/cfre/cfrerr.nls:
```diff
@@ -1590,3 +1590,79 @@ J9NLS_CFR_ERR_PREVIEW_VERSION2.explanation=Classfiles with minor version 65535 m
+J9NLS_CFR_ERR_ENCLOSING_METHOD_CLASS_INDEX_IS_REFERENCE_TYPE.system_action=The JVM will throw a verification or classloading-related exception such as java.lang.ClassFormatError.
+J9NLS_CFR_ERR_ENCLOSING_METHOD_CLASS_INDEX_IS_REFERENCE_TYPE.user_response=Contact the provider of the classfile for a corrected version.
+
+NEST_HOST_INVALID_REFERENCE_TYPE_DESCRIPTOR=The host class index of a nethost attribute cannot be a referencetype descriptor
```
"nethost" --> "nesthost"

hangshao0 replied:
Better to use "The CONSTANT_Class_info at the host_class_index cannot be a reference type descriptor". And add
 `# CONSTANT_Class_info is not translatable` 
 `# host_class_index is not translatable` 

above the message.

hangshao0 commented on runtime/nls/cfre/cfrerr.nls:
```diff
@@ -1590,3 +1590,79 @@ J9NLS_CFR_ERR_PREVIEW_VERSION2.explanation=Classfiles with minor version 65535 m
+NEST_HOST_INVALID_REFERENCE_TYPE_DESCRIPTOR.system_action=The JVM will throw a verification or classloading-related exception such as java.lang.ClassFormatError.
+NEST_HOST_INVALID_REFERENCE_TYPE_DESCRIPTOR.user_response=Contact the provider of the classfile for a corrected version.
+
+NEW_INVALID_REFERENCE_TYPE_DESCRIPTOR=The new instruction cannot operate on a referencetype descriptor
```
Please add

`#new is not translatable`

hangshao0 commented on runtime/nls/cfre/cfrerr.nls:
```diff
@@ -1590,3 +1590,79 @@ J9NLS_CFR_ERR_PREVIEW_VERSION2.explanation=Classfiles with minor version 65535 m
+J9NLS_CFR_ERR_EXCEPTION_IS_REFERENCE_TYPE_DESCRIPTOR.system_action=The JVM will throw a verification or classloading-related exception such as java.lang.ClassFormatError.
+J9NLS_CFR_ERR_EXCEPTION_IS_REFERENCE_TYPE_DESCRIPTOR.user_response=Contact the provider of the classfile for a corrected version.
+
+J9NLS_CFR_ERR_INNER_CLASS_REFERENCE_TYPE=Inner class info cannot be a referencetype descriptor
```
Better to use "The CONSTANT_Class_info at the inner_class_info_index cannot be a reference type descriptor". And add 
`# CONSTANT_Class_info is not translatable`
` # inner_class_info_index is not translatable`
above the message.

hangshao0 commented on runtime/nls/cfre/cfrerr.nls:
```diff
@@ -1590,3 +1590,79 @@ J9NLS_CFR_ERR_PREVIEW_VERSION2.explanation=Classfiles with minor version 65535 m
+# START NON-TRANSLATABLE
+NEST_HOST_INVALID_REFERENCE_TYPE_DESCRIPTOR.explanation=Please consult the Java Virtual Machine Specification for a detailed explanation.
+NEST_HOST_INVALID_REFERENCE_TYPE_DESCRIPTOR.system_action=The JVM will throw a verification or classloading-related exception such as java.lang.ClassFormatError.
+NEST_HOST_INVALID_REFERENCE_TYPE_DESCRIPTOR.user_response=Contact the provider of the classfile for a corrected version.
```
Please add an empty line at the end of the file.

hangshao0 commented on runtime/nls/cfre/cfrerr.nls:
```diff
@@ -1590,3 +1590,79 @@ J9NLS_CFR_ERR_PREVIEW_VERSION2.explanation=Classfiles with minor version 65535 m
+NEW_INVALID_REFERENCE_TYPE_DESCRIPTOR.system_action=The JVM will throw a verification or classloading-related exception such as java.lang.ClassFormatError.
+NEW_INVALID_REFERENCE_TYPE_DESCRIPTOR.user_response=Contact the provider of the classfile for a corrected version.
+
+ACONST_INIT_INVALID_ARRAY=The aconst_init instruction cannot operate on an array
```
Please add

`# aconst_init is not translatable`

hangshao0 commented on runtime/verbose/errormessageframeworkrtv.c:
```diff
@@ -435,6 +435,12 @@ getBCVDataType(J9BytecodeVerificationData* verifyData, MethodContextInfo* method
 		typeNameIndex = CFR_STACKMAP_TYPE_OBJECT;
 
+#if defined(J9VM_OPT_VALHALLA_VALUE_TYPES)
+		if (bcvType & BCV_PRIMITIVE_VALUETYPE) {
```
Please use `J9_ARE_ALL_BITS_SET()`

hangshao0 commented on runtime/verbose/errormessagehelper.c:
```diff
@@ -629,8 +629,16 @@ convertBcvToCfrType(MethodContextInfo* methodInfo, StackMapFrame* stackMapFrame,
+		{
+			UDATA objTypeTag = CFR_STACKMAP_TYPE_OBJECT;
+	#if defined(J9VM_OPT_VALHALLA_VALUE_TYPES)
+			if (bcvType & BCV_PRIMITIVE_VALUETYPE) {
```
Please use J9_ARE_ALL_BITS_SET()

hangshao0 commented on runtime/oti/cfreader.h:
```diff
@@ -75,6 +75,9 @@
 #define	CFR_STACKMAP_TYPE_BOOL_ARRAY			0x10
+#if defined(J9VM_OPT_VALHALLA_VALUE_TYPES)
+#define CFR_STACKMAP_TYPE_PRIMITIVE_OBJECT		0x11
+#endif /* #if defined(J9VM_OPT_VALHALLA_NEW_FACTORY_METHOD) */
```
The comment is incorrect, Should be `/* defined(J9VM_OPT_VALHALLA_VALUE_TYPES) */`

hangshao0 commented on runtime/bcverify/vrfyhelp.c:
```diff
@@ -748,9 +764,9 @@ isClassCompatibleByName(J9BytecodeVerificationData *verifyData, UDATA sourceClas
-	if( sourceClass == BCV_BASE_TYPE_NULL ) 
-		return (IDATA) TRUE;
+	/* NULL is compatible with non Q type classes */
+	if(sourceClass == BCV_BASE_TYPE_NULL)
```
Please put the `return` statement in `{ }` .

ehrenjulzert replied:
addressed this and your other comments in a2d8527

tajila commented on runtime/bcverify/staticverify.c:
```diff
@@ -1802,10 +1816,12 @@ j9bcv_verifyClassStructure (J9PortLibrary * portLib, J9CfrClassFile * classfile,
 							invalidRetType = TRUE;
 						}
 					}
+#if !defined(J9VM_OPT_VALHALLA_VALUE_TYPES) /* This #if !defined check should be removed (along with all other code that references J9VM_OPT_VALHALLA_NEW_FACTORY_METHOD) once the <vnew> method is introduced */
```
is there an issue for this?

Also, can you please put the comment above the macro

tajila commented on runtime/bcverify/vrfyhelp.c:
```diff
@@ -256,8 +256,16 @@ buildStackFromMethodSignature( J9BytecodeVerificationData *verifyData, UDATA **s
 			/* This is <init>, not java/lang/Object */
-			PUSH(BCV_SPECIAL_INIT | (classIndex << BCV_CLASS_INDEX_SHIFT));
+#if defined(J9VM_OPT_VALHALLA_VALUE_TYPES)
+			if (J9ROMCLASS_IS_PRIMITIVE_VALUE_TYPE(romClass)) { /* This is temporary and will have to be changed when <vnew> is introduced (along with all other code that references J9VM_OPT_VALHALLA_NEW_FACTORY_METHOD) */
```
same here

tajila commented on runtime/verbose/errormessagehelper.c:
```diff
@@ -978,6 +989,16 @@ printTypeInfoToBuffer(MessageBuffer* buf, U_8 tag, J9UTF8Ref* dataType, BOOLEAN
+		/* Object, init/new object and object 'reference' */
+		printMessage(buf, "'%.*s%.*s%.*s'",
+				 1, "Q",	/* class prefix 'Q' for the primitive object */
+				dataType->length, dataType->bytes,	/* class name		*/
```
formatting with `/* class name		*/`

tajila commented on runtime/verutil/chverify.c:
```diff
@@ -212,3 +212,30 @@ I_32
+		/* All referencetype descriptors must be more than one character long */
+		result = FALSE;
+	} else if (
+		('[' == bytes[0]) || /* Check if bytes is an ArrayType descriptor */
```
formatting:

Should be more lile

```
else if (condition1
    || condition2
) {
   ...
}
```

tajila commented on runtime/bcverify/vrfyhelp.c:
```diff
@@ -759,7 +776,8 @@ isClassCompatibleByName(J9BytecodeVerificationData *verifyData, UDATA sourceClas
 		index = parseObjectOrArrayName(verifyData, targetClassName);
 	} else {
-		index = convertClassNameToStackMapType(verifyData, targetClassName, (U_16)targetClassNameLength, 0, 0);
+		UDATA type = IS_QTYPE(*targetClassName) ? BCV_PRIMITIVE_VALUETYPE : BCV_OBJECT_OR_ARRAY;
```
We typically dont use `? : ` outside of macros

You can add:

`#define BCV_GET_TYPE_FROM_CHAR(char) ...`

ehrenjulzert replied:
addressed this and your other comments in d010fce

tajila commented on runtime/bcutil/cfreader.c:
```diff
@@ -1488,10 +1488,11 @@ checkPool(J9CfrClassFile* classfile, U_8* segment, U_8* poolStart, I_32 *maxBoot
 				goto _errorFound;
 			}
-			if (cpBase[info->slot2].tag != CFR_CONSTANT_Utf8) {
+			utf8 = &cpBase[info->slot2];
```
why was this change made?

tajila commented on runtime/bcutil/cfreader.c:
```diff
@@ -1501,7 +1502,8 @@ checkPool(J9CfrClassFile* classfile, U_8* segment, U_8* poolStart, I_32 *maxBoot
 				goto _errorFound;
 			}
-			if (cpBase[info->slot1].tag != CFR_CONSTANT_Utf8) {
+			utf8 = &cpBase[info->slot1];
```
same here

ehrenjulzert replied:
I had some written some code in these 2 places that made use of the new `utf8` variable, but that code has now been removed. I guess I should remove the `utf8` variable as well and revert the code to how it was previously?

tajila replied:
You can just revert the code

ehrenjulzert commented on runtime/bcverify/vrfyhelp.c:
```diff
@@ -256,8 +256,17 @@ buildStackFromMethodSignature( J9BytecodeVerificationData *verifyData, UDATA **s
+				PUSH(BCV_SPECIAL_INIT | (classIndex << BCV_CLASS_INDEX_SHIFT));
+#if defined(J9VM_OPT_VALHALLA_VALUE_TYPES)
+			}
 			isUninitializedThis = TRUE;
```
Good catch, I meant to include that in all builds

tajila replied:
you could rewrite it as:

```
#if defined(J9VM_OPT_VALHALLA_VALUE_TYPES)
			/* This is temporary and will have to be changed when <vnew> is introduced (along with all other code that references J9VM_OPT_VALHALLA_NEW_FACTORY_METHOD) */
			if (J9ROMCLASS_IS_PRIMITIVE_VALUE_TYPE(romClass)) {
				PUSH(BCV_PRIMITIVE_VALUETYPE | BCV_SPECIAL_INIT | (classIndex << BCV_CLASS_INDEX_SHIFT));
			} else 
#endif /* #if defined(J9VM_OPT_VALHALLA_VALUE_TYPES) */
                       {
                              PUSH(BCV_SPECIAL_INIT | (classIndex << BCV_CLASS_INDEX_SHIFT));
                       }
```

that way you dont need the macro on the last `}`

tajila commented on runtime/oti/bytecodewalk.h:
```diff
@@ -102,6 +102,11 @@ base types: (in the 19bit class index field)
+#define BCV_PRIMITIVE_VALUETYPE				16		/* set bit means primitive class */
+#endif /* #if defined(J9VM_OPT_VALHALLA_VALUE_TYPES) */
+
+#define BCV_GET_TYPE_FROM_CHAR(char) (IS_QTYPE(char) ? BCV_PRIMITIVE_VALUETYPE : BCV_OBJECT_OR_ARRAY)
```
should be IS_QTYPE((char))

ehrenjulzert replied:
addressed in 0c91935 along with your other comments. I also defined `BCV_GET_TYPE_FROM_CHAR(char)` as just `(BCV_OBJECT_OR_ARRAY)` in non-VT builds because `BCV_PRIMITIVE_VALUETYPE` isn't defined in that case.

tajila commented on runtime/oti/bytecodewalk.h:
```diff
@@ -102,6 +102,15 @@ base types: (in the 19bit class index field)
+#endif /* #if defined(J9VM_OPT_VALHALLA_VALUE_TYPES) */
+
+#if defined(J9VM_OPT_VALHALLA_VALUE_TYPES)
+#define BCV_GET_TYPE_FROM_CHAR(char) (IS_QTYPE((char)) ? BCV_PRIMITIVE_VALUETYPE : BCV_OBJECT_OR_ARRAY)
```
You can define `BCV_PRIMITIVE_VALUETYPE` unconditionally, the compiler ignores constants that arent used and I dont think we need two versions of `BCV_GET_TYPE_FROM_CHAR`

ehrenjulzert replied:
Updated in 3116a40

ehrenjulzert commented:
Just rebased to keep this branch up to date

tajila commented:
Jenkins test sanity,extended plinuxval jdknext

tajila commented:
Jenkins test sanity win jdk8

tajila commented:
Jenkins test sanity xlinuxvalst jdknext

ehrenjulzert commented:
Windows build is failing because of a UDATA being used instead of a U_8, fixing it now

hangshao0 commented:
I expect @ehrenjulzert will push shortly for the Windows failure. Making this comment to save the link of running build: https://openj9-jenkins.osuosl.org/job/Pipeline_Build_Test_JDKnext_x86-64_linux_vt_standard/39/

tajila commented:
Jenkins test sanity win jdk8

tajila commented:
Jenkins test sanity plinuxvalst jdknext

tajila commented:
Jenkins test sanity xlinuxvalst jdknext

tajila commented:
There is a hang in ` JCL_TEST_Test-Annotation-Package`

hangshao0 commented:
@ehrenjulzert Did you figure out the cause of the failure in the PR builds ?

JasonFengJ9 commented:
https://openj9-jenkins.osuosl.org/job/Test_openjdknext_j9_sanity.functional_ppc64le_linux_vt_standard_Personal_testList_0/13/console
```
10:21:14  [IncludeExcludeTestAnnotationTransformer] [INFO] EXCLUDE_FILE environment variable: /home/jenkins/workspace/Test_openjdknext_j9_sanity.functional_ppc64le_linux_vt_standard_Personal_testList_0/aqa-tests/TKG/../TestConfig/resources/excludes/latest_exclude_20.txt,/home/jenkins/workspace/Test_openjdknext_j9_sanity.functional_ppc64le_linux_vt_standard_Personal_testList_0/aqa-tests/TKG/../TestConfig/resources/excludes/feature_ojdkmh_exclude.txt,/home/jenkins/workspace/Test_openjdknext_j9_sanity.functional_ppc64le_linux_vt_standard_Personal_testList_0/aqa-tests/TKG/../TestConfig/resources/excludes/feature_vtstandard_exclude.txt
10:21:14  [IncludeExcludeTestAnnotationTransformer] [INFO] Unable to open file 
```
It was looking for `latest_exclude_20.txt` instead of `latest_exclude_19.txt` which excluded a few known project loom related tests.
Will add `latest_exclude_20.txt`.

fyi @ehrenjulzert

keithc-ca commented on runtime/nls/cfre/cfrerr.nls:
```diff
@@ -1590,3 +1590,98 @@ J9NLS_CFR_ERR_PREVIEW_VERSION2.explanation=Classfiles with minor version 65535 m
+# END NON-TRANSLATABLE
+
+# ReferenceType is not translatable
+NEST_MEMBER_INVALID_REFERENCETYPE_DESCRIPTOR=Nest member classes cannot be ReferenceType descriptors
```
Should this not be `J9NLS_CFR_ERR_NEST_HOST_INVALID_REFERENCETYPE_DESCRIPTOR` for consistency with other keys and the auxiliary information below?

ehrenjulzert replied:
Good point, thanks for the PR
